// z3-solver/build/high-level/high-level.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createApi = void 0;
// TODO(ritave): Add typing for Context Options
//               https://github.com/Z3Prover/z3/pull/6048#discussion_r883391669
// TODO(ritave): Add an error handler
// TODO(ritave): Add support for building faster floats without support for Safari
// TODO(ritave): Use Z3_DECLARE_CLOSURE macro to generate code https://github.com/Z3Prover/z3/pull/6048#discussion_r884155462
// TODO(ritave): Add pretty printing
// TODO(ritave): Make Z3 multi-threaded
// TODO(ritave): If a test times out, jest kills it, and the global state of Z3 is left in an unexpected state.
//               This occurs specifically during longer check(). Afterwards, all next tests will fail to run
//               thinking the previous call was not finished. Find a way to stop execution and clean up the global state
const async_mutex_1 = require("async-mutex");
const low_level_1 = require("../low-level");
const types_1 = require("./types");
const utils_1 = require("./utils");
const FALLBACK_PRECISION = 17;
const asyncMutex = new async_mutex_1.Mutex();
function isCoercibleRational(obj) {
    // prettier-ignore
    const r = ((obj !== null &&
        (typeof obj === 'object' || typeof obj === 'function')) &&
        (obj.numerator !== null &&
            (typeof obj.numerator === 'number' || typeof obj.numerator === 'bigint')) &&
        (obj.denominator !== null &&
            (typeof obj.denominator === 'number' || typeof obj.denominator === 'bigint')));
    r &&
        (0, utils_1.assert)((typeof obj.numerator !== 'number' || Number.isSafeInteger(obj.numerator)) &&
            (typeof obj.denominator !== 'number' || Number.isSafeInteger(obj.denominator)), 'Fraction numerator and denominator must be integers');
    return r;
}
function createApi(Z3) {
    // TODO(ritave): Create a custom linting rule that checks if the provided callbacks to cleanup
    //               Don't capture `this`
    const cleanup = new FinalizationRegistry(callback => callback());
    function enableTrace(tag) {
        Z3.enable_trace(tag);
    }
    function disableTrace(tag) {
        Z3.disable_trace(tag);
    }
    function getVersion() {
        return Z3.get_version();
    }
    function getVersionString() {
        const { major, minor, build_number } = Z3.get_version();
        return `${major}.${minor}.${build_number}`;
    }
    function getFullVersion() {
        return Z3.get_full_version();
    }
    function openLog(filename) {
        return Z3.open_log(filename);
    }
    function appendLog(s) {
        Z3.append_log(s);
    }
    function setParam(key, value) {
        if (typeof key === 'string') {
            Z3.global_param_set(key, value.toString());
        }
        else {
            (0, utils_1.assert)(value === undefined, "Can't provide a Record and second parameter to set_param at the same time");
            Object.entries(key).forEach(([key, value]) => setParam(key, value));
        }
    }
    function resetParams() {
        Z3.global_param_reset_all();
    }
    function getParam(name) {
        return Z3.global_param_get(name);
    }
    function createContext(name, options) {
        const cfg = Z3.mk_config();
        if (options != null) {
            Object.entries(options).forEach(([key, value]) => check(Z3.set_param_value(cfg, key, value.toString())));
        }
        const contextPtr = Z3.mk_context_rc(cfg);
        Z3.set_ast_print_mode(contextPtr, low_level_1.Z3_ast_print_mode.Z3_PRINT_SMTLIB2_COMPLIANT);
        Z3.del_config(cfg);
        function _assertContext(...ctxs) {
            ctxs.forEach(other => (0, utils_1.assert)('ctx' in other ? ctx === other.ctx : ctx === other, 'Context mismatch'));
        }
        // call this after every nontrivial call to the underlying API
        function throwIfError() {
            if (Z3.get_error_code(contextPtr) !== low_level_1.Z3_error_code.Z3_OK) {
                throw new Error(Z3.get_error_msg(ctx.ptr, Z3.get_error_code(ctx.ptr)));
            }
        }
        function check(val) {
            throwIfError();
            return val;
        }
        /////////////
        // Private //
        /////////////
        function _toSymbol(s) {
            if (typeof s === 'number') {
                return check(Z3.mk_int_symbol(contextPtr, s));
            }
            else {
                return check(Z3.mk_string_symbol(contextPtr, s));
            }
        }
        function _fromSymbol(sym) {
            const kind = check(Z3.get_symbol_kind(contextPtr, sym));
            switch (kind) {
                case low_level_1.Z3_symbol_kind.Z3_INT_SYMBOL:
                    return Z3.get_symbol_int(contextPtr, sym);
                case low_level_1.Z3_symbol_kind.Z3_STRING_SYMBOL:
                    return Z3.get_symbol_string(contextPtr, sym);
                default:
                    (0, utils_1.assertExhaustive)(kind);
            }
        }
        function _toAst(ast) {
            switch (check(Z3.get_ast_kind(contextPtr, ast))) {
                case low_level_1.Z3_ast_kind.Z3_SORT_AST:
                    return _toSort(ast);
                case low_level_1.Z3_ast_kind.Z3_FUNC_DECL_AST:
                    return new FuncDeclImpl(ast);
                default:
                    return _toExpr(ast);
            }
        }
        function _toSort(ast) {
            switch (check(Z3.get_sort_kind(contextPtr, ast))) {
                case low_level_1.Z3_sort_kind.Z3_BOOL_SORT:
                    return new BoolSortImpl(ast);
                case low_level_1.Z3_sort_kind.Z3_INT_SORT:
                case low_level_1.Z3_sort_kind.Z3_REAL_SORT:
                    return new ArithSortImpl(ast);
                case low_level_1.Z3_sort_kind.Z3_BV_SORT:
                    return new BitVecSortImpl(ast);
                default:
                    return new SortImpl(ast);
            }
        }
        function _toExpr(ast) {
            const kind = check(Z3.get_ast_kind(contextPtr, ast));
            if (kind === low_level_1.Z3_ast_kind.Z3_QUANTIFIER_AST) {
                (0, utils_1.assert)(false);
            }
            const sortKind = check(Z3.get_sort_kind(contextPtr, Z3.get_sort(contextPtr, ast)));
            switch (sortKind) {
                case low_level_1.Z3_sort_kind.Z3_BOOL_SORT:
                    return new BoolImpl(ast);
                case low_level_1.Z3_sort_kind.Z3_INT_SORT:
                    if (kind === low_level_1.Z3_ast_kind.Z3_NUMERAL_AST) {
                        return new IntNumImpl(ast);
                    }
                    return new ArithImpl(ast);
                case low_level_1.Z3_sort_kind.Z3_REAL_SORT:
                    if (kind === low_level_1.Z3_ast_kind.Z3_NUMERAL_AST) {
                        return new RatNumImpl(ast);
                    }
                    return new ArithImpl(ast);
                case low_level_1.Z3_sort_kind.Z3_BV_SORT:
                    if (kind === low_level_1.Z3_ast_kind.Z3_NUMERAL_AST) {
                        return new BitVecNumImpl(ast);
                    }
                    return new BitVecImpl(ast);
                default:
                    return new ExprImpl(ast);
            }
        }
        function _flattenArgs(args) {
            const result = [];
            for (const arg of args) {
                if (isAstVector(arg)) {
                    result.push(...arg.values());
                }
                else {
                    result.push(arg);
                }
            }
            return result;
        }
        function _toProbe(p) {
            if (isProbe(p)) {
                return p;
            }
            return new ProbeImpl(p);
        }
        function _probeNary(f, args) {
            (0, utils_1.assert)(args.length > 0, 'At least one argument expected');
            let r = _toProbe(args[0]);
            for (let i = 1; i < args.length; i++) {
                r = new ProbeImpl(check(f(contextPtr, r.ptr, _toProbe(args[i]).ptr)));
            }
            return r;
        }
        ///////////////
        // Functions //
        ///////////////
        function interrupt() {
            check(Z3.interrupt(contextPtr));
        }
        function isModel(obj) {
            const r = obj instanceof ModelImpl;
            r && _assertContext(obj);
            return r;
        }
        function isAst(obj) {
            const r = obj instanceof AstImpl;
            r && _assertContext(obj);
            return r;
        }
        function isSort(obj) {
            const r = obj instanceof SortImpl;
            r && _assertContext(obj);
            return r;
        }
        function isFuncDecl(obj) {
            const r = obj instanceof FuncDeclImpl;
            r && _assertContext(obj);
            return r;
        }
        function isApp(obj) {
            if (!isExpr(obj)) {
                return false;
            }
            const kind = check(Z3.get_ast_kind(contextPtr, obj.ast));
            return kind === low_level_1.Z3_ast_kind.Z3_NUMERAL_AST || kind === low_level_1.Z3_ast_kind.Z3_APP_AST;
        }
        function isConst(obj) {
            return isExpr(obj) && isApp(obj) && obj.numArgs() === 0;
        }
        function isExpr(obj) {
            const r = obj instanceof ExprImpl;
            r && _assertContext(obj);
            return r;
        }
        function isVar(obj) {
            return isExpr(obj) && check(Z3.get_ast_kind(contextPtr, obj.ast)) === low_level_1.Z3_ast_kind.Z3_VAR_AST;
        }
        function isAppOf(obj, kind) {
            return isExpr(obj) && isApp(obj) && obj.decl().kind() === kind;
        }
        function isBool(obj) {
            const r = obj instanceof BoolImpl;
            r && _assertContext(obj);
            return r;
        }
        function isTrue(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_TRUE);
        }
        function isFalse(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_FALSE);
        }
        function isAnd(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_AND);
        }
        function isOr(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_OR);
        }
        function isImplies(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_IMPLIES);
        }
        function isNot(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_NOT);
        }
        function isEq(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_EQ);
        }
        function isDistinct(obj) {
            return isAppOf(obj, low_level_1.Z3_decl_kind.Z3_OP_DISTINCT);
        }
        function isArith(obj) {
            const r = obj instanceof ArithImpl;
            r && _assertContext(obj);
            return r;
        }
        function isArithSort(obj) {
            const r = obj instanceof ArithSortImpl;
            r && _assertContext(obj);
            return r;
        }
        function isInt(obj) {
            return isArith(obj) && isIntSort(obj.sort);
        }
        function isIntVal(obj) {
            const r = obj instanceof IntNumImpl;
            r && _assertContext(obj);
            return r;
        }
        function isIntSort(obj) {
            return isSort(obj) && obj.kind() === low_level_1.Z3_sort_kind.Z3_INT_SORT;
        }
        function isReal(obj) {
            return isArith(obj) && isRealSort(obj.sort);
        }
        function isRealVal(obj) {
            const r = obj instanceof RatNumImpl;
            r && _assertContext(obj);
            return r;
        }
        function isRealSort(obj) {
            return isSort(obj) && obj.kind() === low_level_1.Z3_sort_kind.Z3_REAL_SORT;
        }
        function isBitVecSort(obj) {
            const r = obj instanceof BitVecSortImpl;
            r && _assertContext(obj);
            return r;
        }
        function isBitVec(obj) {
            const r = obj instanceof BitVecImpl;
            r && _assertContext(obj);
            return r;
        }
        function isBitVecVal(obj) {
            const r = obj instanceof BitVecNumImpl;
            r && _assertContext(obj);
            return r;
        }
        function isProbe(obj) {
            const r = obj instanceof ProbeImpl;
            r && _assertContext(obj);
            return r;
        }
        function isTactic(obj) {
            const r = obj instanceof TacticImpl;
            r && _assertContext(obj);
            return r;
        }
        function isAstVector(obj) {
            const r = obj instanceof AstVectorImpl;
            r && _assertContext(obj);
            return r;
        }
        function eqIdentity(a, b) {
            return a.eqIdentity(b);
        }
        function getVarIndex(obj) {
            (0, utils_1.assert)(isVar(obj), 'Z3 bound variable expected');
            return Z3.get_index_value(contextPtr, obj.ast);
        }
        function from(value) {
            if (typeof value === 'boolean') {
                return Bool.val(value);
            }
            else if (typeof value === 'number') {
                if (!Number.isFinite(value)) {
                    throw new Error(`cannot represent infinity/NaN (got ${value})`);
                }
                if (Math.floor(value) === value) {
                    return Int.val(value);
                }
                return Real.val(value);
            }
            else if (isCoercibleRational(value)) {
                return Real.val(value);
            }
            else if (typeof value === 'bigint') {
                return Int.val(value);
            }
            else if (isExpr(value)) {
                return value;
            }
            (0, utils_1.assert)(false);
        }
        async function solve(...assertions) {
            const solver = new ctx.Solver();
            solver.add(...assertions);
            const result = await solver.check();
            if (result === 'sat') {
                return solver.model();
            }
            return result;
        }
        ///////////////////////////////
        // expression simplification //
        ///////////////////////////////
        async function simplify(e) {
            const result = await Z3.simplify(contextPtr, e.ast);
            return _toExpr(check(result));
        }
        /////////////
        // Objects //
        /////////////
        const Sort = {
            declare: (name) => new SortImpl(Z3.mk_uninterpreted_sort(contextPtr, _toSymbol(name))),
        };
        const Function = {
            declare: (name, ...signature) => {
                const arity = signature.length - 1;
                const rng = signature[arity];
                _assertContext(rng);
                const dom = [];
                for (let i = 0; i < arity; i++) {
                    _assertContext(signature[i]);
                    dom.push(signature[i].ptr);
                }
                return new FuncDeclImpl(Z3.mk_func_decl(contextPtr, _toSymbol(name), dom, rng.ptr));
            },
            fresh: (...signature) => {
                const arity = signature.length - 1;
                const rng = signature[arity];
                _assertContext(rng);
                const dom = [];
                for (let i = 0; i < arity; i++) {
                    _assertContext(signature[i]);
                    dom.push(signature[i].ptr);
                }
                return new FuncDeclImpl(Z3.mk_fresh_func_decl(contextPtr, 'f', dom, rng.ptr));
            },
        };
        const RecFunc = {
            declare: (name, ...signature) => {
                const arity = signature.length - 1;
                const rng = signature[arity];
                _assertContext(rng);
                const dom = [];
                for (let i = 0; i < arity; i++) {
                    _assertContext(signature[i]);
                    dom.push(signature[i].ptr);
                }
                return new FuncDeclImpl(Z3.mk_rec_func_decl(contextPtr, _toSymbol(name), dom, rng.ptr));
            },
            addDefinition: (f, args, body) => {
                _assertContext(f, ...args, body);
                check(Z3.add_rec_def(contextPtr, f.ptr, args.map(arg => arg.ast), body.ast));
            },
        };
        const Bool = {
            sort: () => new BoolSortImpl(Z3.mk_bool_sort(contextPtr)),
            const: (name) => new BoolImpl(Z3.mk_const(contextPtr, _toSymbol(name), Bool.sort().ptr)),
            consts: (names) => {
                if (typeof names === 'string') {
                    names = names.split(' ');
                }
                return names.map(name => Bool.const(name));
            },
            vector: (prefix, count) => {
                const result = [];
                for (let i = 0; i < count; i++) {
                    result.push(Bool.const(`${prefix}__${i}`));
                }
                return result;
            },
            fresh: (prefix = 'b') => new BoolImpl(Z3.mk_fresh_const(contextPtr, prefix, Bool.sort().ptr)),
            val: (value) => {
                if (value) {
                    return new BoolImpl(Z3.mk_true(contextPtr));
                }
                return new BoolImpl(Z3.mk_false(contextPtr));
            },
        };
        const Int = {
            sort: () => new ArithSortImpl(Z3.mk_int_sort(contextPtr)),
            const: (name) => new ArithImpl(Z3.mk_const(contextPtr, _toSymbol(name), Int.sort().ptr)),
            consts: (names) => {
                if (typeof names === 'string') {
                    names = names.split(' ');
                }
                return names.map(name => Int.const(name));
            },
            vector: (prefix, count) => {
                const result = [];
                for (let i = 0; i < count; i++) {
                    result.push(Int.const(`${prefix}__${i}`));
                }
                return result;
            },
            fresh: (prefix = 'x') => new ArithImpl(Z3.mk_fresh_const(contextPtr, prefix, Int.sort().ptr)),
            val: (value) => {
                (0, utils_1.assert)(typeof value === 'bigint' || typeof value === 'string' || Number.isSafeInteger(value));
                return new IntNumImpl(check(Z3.mk_numeral(contextPtr, value.toString(), Int.sort().ptr)));
            },
        };
        const Real = {
            sort: () => new ArithSortImpl(Z3.mk_real_sort(contextPtr)),
            const: (name) => new ArithImpl(check(Z3.mk_const(contextPtr, _toSymbol(name), Real.sort().ptr))),
            consts: (names) => {
                if (typeof names === 'string') {
                    names = names.split(' ');
                }
                return names.map(name => Real.const(name));
            },
            vector: (prefix, count) => {
                const result = [];
                for (let i = 0; i < count; i++) {
                    result.push(Real.const(`${prefix}__${i}`));
                }
                return result;
            },
            fresh: (prefix = 'b') => new ArithImpl(Z3.mk_fresh_const(contextPtr, prefix, Real.sort().ptr)),
            val: (value) => {
                if (isCoercibleRational(value)) {
                    value = `${value.numerator}/${value.denominator}`;
                }
                return new RatNumImpl(Z3.mk_numeral(contextPtr, value.toString(), Real.sort().ptr));
            },
        };
        const BitVec = {
            sort(bits) {
                (0, utils_1.assert)(Number.isSafeInteger(bits), 'number of bits must be an integer');
                return new BitVecSortImpl(Z3.mk_bv_sort(contextPtr, bits));
            },
            const(name, bits) {
                return new BitVecImpl(check(Z3.mk_const(contextPtr, _toSymbol(name), isBitVecSort(bits) ? bits.ptr : BitVec.sort(bits).ptr)));
            },
            consts(names, bits) {
                if (typeof names === 'string') {
                    names = names.split(' ');
                }
                return names.map(name => BitVec.const(name, bits));
            },
            val(value, bits) {
                if (value === true) {
                    return BitVec.val(1, bits);
                }
                else if (value === false) {
                    return BitVec.val(0, bits);
                }
                return new BitVecNumImpl(check(Z3.mk_numeral(contextPtr, value.toString(), isBitVecSort(bits) ? bits.ptr : BitVec.sort(bits).ptr)));
            },
        };
        function If(condition, onTrue, onFalse) {
            if (isProbe(condition) && isTactic(onTrue) && isTactic(onFalse)) {
                return Cond(condition, onTrue, onFalse);
            }
            (0, utils_1.assert)(!isProbe(condition) && !isTactic(onTrue) && !isTactic(onFalse), 'Mixed expressions and goals');
            if (typeof condition === 'boolean') {
                condition = Bool.val(condition);
            }
            onTrue = from(onTrue);
            onFalse = from(onFalse);
            return _toExpr(check(Z3.mk_ite(contextPtr, condition.ptr, onTrue.ast, onFalse.ast)));
        }
        function Distinct(...exprs) {
            (0, utils_1.assert)(exprs.length > 0, "Can't make Distinct ouf of nothing");
            return new BoolImpl(check(Z3.mk_distinct(contextPtr, exprs.map(expr => {
                expr = from(expr);
                _assertContext(expr);
                return expr.ast;
            }))));
        }
        function Const(name, sort) {
            _assertContext(sort);
            return _toExpr(check(Z3.mk_const(contextPtr, _toSymbol(name), sort.ptr)));
        }
        function Consts(names, sort) {
            _assertContext(sort);
            if (typeof names === 'string') {
                names = names.split(' ');
            }
            return names.map(name => Const(name, sort));
        }
        function FreshConst(sort, prefix = 'c') {
            _assertContext(sort);
            return _toExpr(Z3.mk_fresh_const(sort.ctx.ptr, prefix, sort.ptr));
        }
        function Var(idx, sort) {
            _assertContext(sort);
            return _toExpr(Z3.mk_bound(sort.ctx.ptr, idx, sort.ptr));
        }
        function Implies(a, b) {
            a = from(a);
            b = from(b);
            _assertContext(a, b);
            return new BoolImpl(check(Z3.mk_implies(contextPtr, a.ptr, b.ptr)));
        }
        function Eq(a, b) {
            a = from(a);
            b = from(b);
            _assertContext(a, b);
            return a.eq(b);
        }
        function Xor(a, b) {
            a = from(a);
            b = from(b);
            _assertContext(a, b);
            return new BoolImpl(check(Z3.mk_xor(contextPtr, a.ptr, b.ptr)));
        }
        function Not(a) {
            if (typeof a === 'boolean') {
                a = from(a);
            }
            _assertContext(a);
            if (isProbe(a)) {
                return new ProbeImpl(check(Z3.probe_not(contextPtr, a.ptr)));
            }
            return new BoolImpl(check(Z3.mk_not(contextPtr, a.ptr)));
        }
        function And(...args) {
            if (args.length == 1 && args[0] instanceof ctx.AstVector) {
                args = [...args[0].values()];
                (0, utils_1.assert)((0, utils_1.allSatisfy)(args, isBool) ?? true, 'AstVector containing not bools');
            }
            const allProbes = (0, utils_1.allSatisfy)(args, isProbe) ?? false;
            if (allProbes) {
                return _probeNary(Z3.probe_and, args);
            }
            else {
                const castArgs = args.map(from);
                _assertContext(...castArgs);
                return new BoolImpl(check(Z3.mk_and(contextPtr, castArgs.map(arg => arg.ptr))));
            }
        }
        function Or(...args) {
            if (args.length == 1 && args[0] instanceof ctx.AstVector) {
                args = [...args[0].values()];
                (0, utils_1.assert)((0, utils_1.allSatisfy)(args, isBool) ?? true, 'AstVector containing not bools');
            }
            const allProbes = (0, utils_1.allSatisfy)(args, isProbe) ?? false;
            if (allProbes) {
                return _probeNary(Z3.probe_or, args);
            }
            else {
                const castArgs = args.map(from);
                _assertContext(...castArgs);
                return new BoolImpl(check(Z3.mk_or(contextPtr, castArgs.map(arg => arg.ptr))));
            }
        }
        function ToReal(expr) {
            expr = from(expr);
            _assertContext(expr);
            (0, utils_1.assert)(isInt(expr), 'Int expression expected');
            return new ArithImpl(check(Z3.mk_int2real(contextPtr, expr.ast)));
        }
        function ToInt(expr) {
            if (!isExpr(expr)) {
                expr = Real.val(expr);
            }
            _assertContext(expr);
            (0, utils_1.assert)(isReal(expr), 'Real expression expected');
            return new ArithImpl(check(Z3.mk_real2int(contextPtr, expr.ast)));
        }
        function IsInt(expr) {
            if (!isExpr(expr)) {
                expr = Real.val(expr);
            }
            _assertContext(expr);
            (0, utils_1.assert)(isReal(expr), 'Real expression expected');
            return new BoolImpl(check(Z3.mk_is_int(contextPtr, expr.ast)));
        }
        function Sqrt(a) {
            if (!isExpr(a)) {
                a = Real.val(a);
            }
            return a.pow('1/2');
        }
        function Cbrt(a) {
            if (!isExpr(a)) {
                a = Real.val(a);
            }
            return a.pow('1/3');
        }
        function BV2Int(a, isSigned) {
            _assertContext(a);
            return new ArithImpl(check(Z3.mk_bv2int(contextPtr, a.ast, isSigned)));
        }
        function Int2BV(a, bits) {
            if (isArith(a)) {
                (0, utils_1.assert)(isInt(a), 'parameter must be an integer');
            }
            else {
                (0, utils_1.assert)(typeof a !== 'number' || Number.isSafeInteger(a), 'parameter must not have decimal places');
                a = Int.val(a);
            }
            return new BitVecImpl(check(Z3.mk_int2bv(contextPtr, bits, a.ast)));
        }
        function Concat(...bitvecs) {
            _assertContext(...bitvecs);
            return bitvecs.reduce((prev, curr) => new BitVecImpl(check(Z3.mk_concat(contextPtr, prev.ast, curr.ast))));
        }
        function Cond(probe, onTrue, onFalse) {
            _assertContext(probe, onTrue, onFalse);
            return new TacticImpl(check(Z3.tactic_cond(contextPtr, probe.ptr, onTrue.ptr, onFalse.ptr)));
        }
        class AstImpl {
            constructor(ptr) {
                this.ptr = ptr;
                this.ctx = ctx;
                const myAst = this.ast;
                Z3.inc_ref(contextPtr, myAst);
                cleanup.register(this, () => Z3.dec_ref(contextPtr, myAst));
            }
            get ast() {
                return this.ptr;
            }
            id() {
                return Z3.get_ast_id(contextPtr, this.ast);
            }
            eqIdentity(other) {
                _assertContext(other);
                return check(Z3.is_eq_ast(contextPtr, this.ast, other.ast));
            }
            neqIdentity(other) {
                _assertContext(other);
                return !this.eqIdentity(other);
            }
            sexpr() {
                return Z3.ast_to_string(contextPtr, this.ast);
            }
            hash() {
                return Z3.get_ast_hash(contextPtr, this.ast);
            }
            toString() {
                return this.sexpr();
            }
        }
        class SolverImpl {
            constructor(ptr = Z3.mk_solver(contextPtr)) {
                this.ctx = ctx;
                let myPtr;
                if (typeof ptr === 'string') {
                    myPtr = check(Z3.mk_solver_for_logic(contextPtr, _toSymbol(ptr)));
                }
                else {
                    myPtr = ptr;
                }
                this.ptr = myPtr;
                Z3.solver_inc_ref(contextPtr, myPtr);
                cleanup.register(this, () => Z3.solver_dec_ref(contextPtr, myPtr));
            }
            push() {
                Z3.solver_push(contextPtr, this.ptr);
            }
            pop(num = 1) {
                Z3.solver_pop(contextPtr, this.ptr, num);
            }
            numScopes() {
                return Z3.solver_get_num_scopes(contextPtr, this.ptr);
            }
            reset() {
                Z3.solver_reset(contextPtr, this.ptr);
            }
            add(...exprs) {
                _flattenArgs(exprs).forEach(expr => {
                    _assertContext(expr);
                    check(Z3.solver_assert(contextPtr, this.ptr, expr.ast));
                });
            }
            addAndTrack(expr, constant) {
                if (typeof constant === 'string') {
                    constant = Bool.const(constant);
                }
                (0, utils_1.assert)(isConst(constant), 'Provided expression that is not a constant to addAndTrack');
                check(Z3.solver_assert_and_track(contextPtr, this.ptr, expr.ast, constant.ast));
            }
            assertions() {
                return new AstVectorImpl(check(Z3.solver_get_assertions(contextPtr, this.ptr)));
            }
            async check(...exprs) {
                const assumptions = _flattenArgs(exprs).map(expr => {
                    _assertContext(expr);
                    return expr.ast;
                });
                const result = await asyncMutex.runExclusive(() => check(Z3.solver_check_assumptions(contextPtr, this.ptr, assumptions)));
                switch (result) {
                    case low_level_1.Z3_lbool.Z3_L_FALSE:
                        return 'unsat';
                    case low_level_1.Z3_lbool.Z3_L_TRUE:
                        return 'sat';
                    case low_level_1.Z3_lbool.Z3_L_UNDEF:
                        return 'unknown';
                    default:
                        (0, utils_1.assertExhaustive)(result);
                }
            }
            model() {
                return new ModelImpl(check(Z3.solver_get_model(contextPtr, this.ptr)));
            }
            toString() {
                return check(Z3.solver_to_string(contextPtr, this.ptr));
            }
        }
        class ModelImpl {
            constructor(ptr = Z3.mk_model(contextPtr)) {
                this.ptr = ptr;
                this.ctx = ctx;
                Z3.model_inc_ref(contextPtr, ptr);
                cleanup.register(this, () => Z3.model_dec_ref(contextPtr, ptr));
            }
            length() {
                return Z3.model_get_num_consts(contextPtr, this.ptr) + Z3.model_get_num_funcs(contextPtr, this.ptr);
            }
            [Symbol.iterator]() {
                return this.values();
            }
            *entries() {
                const length = this.length();
                for (let i = 0; i < length; i++) {
                    yield [i, this.get(i)];
                }
            }
            *keys() {
                for (const [key] of this.entries()) {
                    yield key;
                }
            }
            *values() {
                for (const [, value] of this.entries()) {
                    yield value;
                }
            }
            decls() {
                return [...this.values()];
            }
            sexpr() {
                return check(Z3.model_to_string(contextPtr, this.ptr));
            }
            toString() {
                return this.sexpr();
            }
            eval(expr, modelCompletion = false) {
                _assertContext(expr);
                const r = check(Z3.model_eval(contextPtr, this.ptr, expr.ast, modelCompletion));
                if (r === null) {
                    throw new types_1.Z3Error('Failed to evaluatio expression in the model');
                }
                return _toExpr(r);
            }
            get(i, to) {
                (0, utils_1.assert)(to === undefined || typeof i === 'number');
                if (typeof i === 'number') {
                    const length = this.length();
                    if (i >= length) {
                        throw new RangeError(`expected index ${i} to be less than length ${length}`);
                    }
                    if (to === undefined) {
                        const numConsts = check(Z3.model_get_num_consts(contextPtr, this.ptr));
                        if (i < numConsts) {
                            return new FuncDeclImpl(check(Z3.model_get_const_decl(contextPtr, this.ptr, i)));
                        }
                        else {
                            return new FuncDeclImpl(check(Z3.model_get_func_decl(contextPtr, this.ptr, i - numConsts)));
                        }
                    }
                    if (to < 0) {
                        to += length;
                    }
                    if (to >= length) {
                        throw new RangeError(`expected index ${to} to be less than length ${length}`);
                    }
                    const result = [];
                    for (let j = i; j < to; j++) {
                        result.push(this.get(j));
                    }
                    return result;
                }
                else if (isFuncDecl(i) || (isExpr(i) && isConst(i))) {
                    const result = this.getInterp(i);
                    (0, utils_1.assert)(result !== null);
                    return result;
                }
                else if (isSort(i)) {
                    return this.getUniverse(i);
                }
                (0, utils_1.assert)(false, 'Number, declaration or constant expected');
            }
            getInterp(expr) {
                (0, utils_1.assert)(isFuncDecl(expr) || isConst(expr), 'Declaration expected');
                if (isConst(expr)) {
                    (0, utils_1.assert)(isExpr(expr));
                    expr = expr.decl();
                }
                (0, utils_1.assert)(isFuncDecl(expr));
                if (expr.arity() === 0) {
                    const result = check(Z3.model_get_const_interp(contextPtr, this.ptr, expr.ptr));
                    if (result === null) {
                        return null;
                    }
                    return _toExpr(result);
                }
                else {
                    const interp = check(Z3.model_get_func_interp(contextPtr, this.ptr, expr.ptr));
                    if (interp === null) {
                        return null;
                    }
                    return new FuncInterpImpl(interp);
                }
            }
            getUniverse(sort) {
                _assertContext(sort);
                return new AstVectorImpl(check(Z3.model_get_sort_universe(contextPtr, this.ptr, sort.ptr)));
            }
        }
        class FuncInterpImpl {
            constructor(ptr) {
                this.ptr = ptr;
                this.ctx = ctx;
                Z3.func_interp_inc_ref(contextPtr, ptr);
                cleanup.register(this, () => Z3.func_interp_dec_ref(contextPtr, ptr));
            }
        }
        class SortImpl extends AstImpl {
            get ast() {
                return Z3.sort_to_ast(contextPtr, this.ptr);
            }
            kind() {
                return Z3.get_sort_kind(contextPtr, this.ptr);
            }
            subsort(other) {
                _assertContext(other);
                return false;
            }
            cast(expr) {
                _assertContext(expr);
                (0, utils_1.assert)(expr.sort.eqIdentity(expr.sort), 'Sort mismatch');
                return expr;
            }
            name() {
                return _fromSymbol(Z3.get_sort_name(contextPtr, this.ptr));
            }
            eqIdentity(other) {
                _assertContext(other);
                return check(Z3.is_eq_sort(contextPtr, this.ptr, other.ptr));
            }
            neqIdentity(other) {
                return !this.eqIdentity(other);
            }
        }
        class FuncDeclImpl extends AstImpl {
            get ast() {
                return Z3.func_decl_to_ast(contextPtr, this.ptr);
            }
            name() {
                return _fromSymbol(Z3.get_decl_name(contextPtr, this.ptr));
            }
            arity() {
                return Z3.get_arity(contextPtr, this.ptr);
            }
            domain(i) {
                (0, utils_1.assert)(i < this.arity(), 'Index out of bounds');
                return _toSort(Z3.get_domain(contextPtr, this.ptr, i));
            }
            range() {
                return _toSort(Z3.get_range(contextPtr, this.ptr));
            }
            kind() {
                return Z3.get_decl_kind(contextPtr, this.ptr);
            }
            params() {
                const n = Z3.get_decl_num_parameters(contextPtr, this.ptr);
                const result = [];
                for (let i = 0; i < n; i++) {
                    const kind = check(Z3.get_decl_parameter_kind(contextPtr, this.ptr, i));
                    switch (kind) {
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_INT:
                            result.push(check(Z3.get_decl_int_parameter(contextPtr, this.ptr, i)));
                            break;
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_DOUBLE:
                            result.push(check(Z3.get_decl_double_parameter(contextPtr, this.ptr, i)));
                            break;
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_RATIONAL:
                            result.push(check(Z3.get_decl_rational_parameter(contextPtr, this.ptr, i)));
                            break;
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_SYMBOL:
                            result.push(check(Z3.get_decl_symbol_parameter(contextPtr, this.ptr, i)));
                            break;
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_SORT:
                            result.push(new SortImpl(check(Z3.get_decl_sort_parameter(contextPtr, this.ptr, i))));
                            break;
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_AST:
                            result.push(new ExprImpl(check(Z3.get_decl_ast_parameter(contextPtr, this.ptr, i))));
                            break;
                        case low_level_1.Z3_parameter_kind.Z3_PARAMETER_FUNC_DECL:
                            result.push(new FuncDeclImpl(check(Z3.get_decl_func_decl_parameter(contextPtr, this.ptr, i))));
                            break;
                        default:
                            (0, utils_1.assertExhaustive)(kind);
                    }
                }
                return result;
            }
            call(...args) {
                (0, utils_1.assert)(args.length === this.arity(), `Incorrect number of arguments to ${this}`);
                return _toExpr(check(Z3.mk_app(contextPtr, this.ptr, args.map((arg, i) => {
                    return this.domain(i).cast(arg).ast;
                }))));
            }
        }
        class ExprImpl extends AstImpl {
            get sort() {
                return _toSort(Z3.get_sort(contextPtr, this.ast));
            }
            eq(other) {
                return new BoolImpl(check(Z3.mk_eq(contextPtr, this.ast, from(other).ast)));
            }
            neq(other) {
                return new BoolImpl(check(Z3.mk_distinct(contextPtr, [this, other].map(expr => from(expr).ast))));
            }
            params() {
                return this.decl().params();
            }
            decl() {
                (0, utils_1.assert)(isApp(this), 'Z3 application expected');
                return new FuncDeclImpl(check(Z3.get_app_decl(contextPtr, check(Z3.to_app(contextPtr, this.ast)))));
            }
            numArgs() {
                (0, utils_1.assert)(isApp(this), 'Z3 applicaiton expected');
                return check(Z3.get_app_num_args(contextPtr, check(Z3.to_app(contextPtr, this.ast))));
            }
            arg(i) {
                (0, utils_1.assert)(isApp(this), 'Z3 applicaiton expected');
                (0, utils_1.assert)(i < this.numArgs(), `Invalid argument index - expected ${i} to be less than ${this.numArgs()}`);
                return _toExpr(check(Z3.get_app_arg(contextPtr, check(Z3.to_app(contextPtr, this.ast)), i)));
            }
            children() {
                const num_args = this.numArgs();
                if (isApp(this)) {
                    const result = [];
                    for (let i = 0; i < num_args; i++) {
                        result.push(this.arg(i));
                    }
                    return result;
                }
                return [];
            }
        }
        class BoolSortImpl extends SortImpl {
            cast(other) {
                if (typeof other === 'boolean') {
                    other = Bool.val(other);
                }
                (0, utils_1.assert)(isExpr(other), 'true, false or Z3 Boolean expression expected.');
                (0, utils_1.assert)(this.eqIdentity(other.sort), 'Value cannot be converted into a Z3 Boolean value');
                return other;
            }
            subsort(other) {
                _assertContext(other.ctx);
                return other instanceof ArithSortImpl;
            }
        }
        class BoolImpl extends ExprImpl {
            not() {
                return Not(this);
            }
            and(other) {
                return And(this, other);
            }
            or(other) {
                return Or(this, other);
            }
            xor(other) {
                return Xor(this, other);
            }
        }
        class ProbeImpl {
            constructor(ptr) {
                this.ptr = ptr;
                this.ctx = ctx;
            }
        }
        class TacticImpl {
            constructor(tactic) {
                this.ctx = ctx;
                let myPtr;
                if (typeof tactic === 'string') {
                    myPtr = check(Z3.mk_tactic(contextPtr, tactic));
                }
                else {
                    myPtr = tactic;
                }
                this.ptr = myPtr;
                Z3.tactic_inc_ref(contextPtr, myPtr);
                cleanup.register(this, () => Z3.tactic_dec_ref(contextPtr, myPtr));
            }
        }
        class ArithSortImpl extends SortImpl {
            cast(other) {
                const sortTypeStr = isIntSort(this) ? 'IntSort' : 'RealSort';
                if (isExpr(other)) {
                    const otherS = other.sort;
                    if (isArith(other)) {
                        if (this.eqIdentity(otherS)) {
                            return other;
                        }
                        else if (isIntSort(otherS) && isRealSort(this)) {
                            return ToReal(other);
                        }
                        (0, utils_1.assert)(false, "Can't cast Real to IntSort without loss");
                    }
                    else if (isBool(other)) {
                        if (isIntSort(this)) {
                            return If(other, 1, 0);
                        }
                        else {
                            return ToReal(If(other, 1, 0));
                        }
                    }
                    (0, utils_1.assert)(false, `Can't cast expression to ${sortTypeStr}`);
                }
                else {
                    if (typeof other !== 'boolean') {
                        if (isIntSort(this)) {
                            (0, utils_1.assert)(!isCoercibleRational(other), "Can't cast fraction to IntSort");
                            return Int.val(other);
                        }
                        return Real.val(other);
                    }
                    (0, utils_1.assert)(false, `Can't cast primitive to ${sortTypeStr}`);
                }
            }
        }
        class ArithImpl extends ExprImpl {
            add(other) {
                return new ArithImpl(check(Z3.mk_add(contextPtr, [this.ast, this.sort.cast(other).ast])));
            }
            mul(other) {
                return new ArithImpl(check(Z3.mk_mul(contextPtr, [this.ast, this.sort.cast(other).ast])));
            }
            sub(other) {
                return new ArithImpl(check(Z3.mk_sub(contextPtr, [this.ast, this.sort.cast(other).ast])));
            }
            pow(exponent) {
                return new ArithImpl(check(Z3.mk_power(contextPtr, this.ast, this.sort.cast(exponent).ast)));
            }
            div(other) {
                return new ArithImpl(check(Z3.mk_div(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            mod(other) {
                return new ArithImpl(check(Z3.mk_mod(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            neg() {
                return new ArithImpl(check(Z3.mk_unary_minus(contextPtr, this.ast)));
            }
            le(other) {
                return new BoolImpl(check(Z3.mk_le(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            lt(other) {
                return new BoolImpl(check(Z3.mk_lt(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            gt(other) {
                return new BoolImpl(check(Z3.mk_gt(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            ge(other) {
                return new BoolImpl(check(Z3.mk_ge(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
        }
        class IntNumImpl extends ArithImpl {
            value() {
                return BigInt(this.asString());
            }
            asString() {
                return Z3.get_numeral_string(contextPtr, this.ast);
            }
            asBinary() {
                return Z3.get_numeral_binary_string(contextPtr, this.ast);
            }
        }
        class RatNumImpl extends ArithImpl {
            value() {
                return { numerator: this.numerator().value(), denominator: this.denominator().value() };
            }
            numerator() {
                return new IntNumImpl(Z3.get_numerator(contextPtr, this.ast));
            }
            denominator() {
                return new IntNumImpl(Z3.get_denominator(contextPtr, this.ast));
            }
            asNumber() {
                const { numerator, denominator } = this.value();
                const div = numerator / denominator;
                return Number(div) + Number(numerator - div * denominator) / Number(denominator);
            }
            asDecimal(prec = Number.parseInt(getParam('precision') ?? FALLBACK_PRECISION.toString())) {
                return Z3.get_numeral_decimal_string(contextPtr, this.ast, prec);
            }
            asString() {
                return Z3.get_numeral_string(contextPtr, this.ast);
            }
        }
        class BitVecSortImpl extends SortImpl {
            size() {
                return Z3.get_bv_sort_size(contextPtr, this.ptr);
            }
            subsort(other) {
                return isBitVecSort(other) && this.size() < other.size();
            }
            cast(other) {
                if (isExpr(other)) {
                    _assertContext(other);
                    return other;
                }
                (0, utils_1.assert)(!isCoercibleRational(other), "Can't convert rational to BitVec");
                return BitVec.val(other, this.size());
            }
        }
        class BitVecImpl extends ExprImpl {
            size() {
                return this.sort.size();
            }
            add(other) {
                return new BitVecImpl(check(Z3.mk_bvadd(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            mul(other) {
                return new BitVecImpl(check(Z3.mk_bvmul(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            sub(other) {
                return new BitVecImpl(check(Z3.mk_bvsub(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            sdiv(other) {
                return new BitVecImpl(check(Z3.mk_bvsdiv(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            udiv(other) {
                return new BitVecImpl(check(Z3.mk_bvudiv(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            smod(other) {
                return new BitVecImpl(check(Z3.mk_bvsmod(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            urem(other) {
                return new BitVecImpl(check(Z3.mk_bvurem(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            srem(other) {
                return new BitVecImpl(check(Z3.mk_bvsrem(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            neg() {
                return new BitVecImpl(check(Z3.mk_bvneg(contextPtr, this.ast)));
            }
            or(other) {
                return new BitVecImpl(check(Z3.mk_bvor(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            and(other) {
                return new BitVecImpl(check(Z3.mk_bvand(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            nand(other) {
                return new BitVecImpl(check(Z3.mk_bvnand(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            xor(other) {
                return new BitVecImpl(check(Z3.mk_bvxor(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            xnor(other) {
                return new BitVecImpl(check(Z3.mk_bvxnor(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            shr(count) {
                return new BitVecImpl(check(Z3.mk_bvashr(contextPtr, this.ast, this.sort.cast(count).ast)));
            }
            lshr(count) {
                return new BitVecImpl(check(Z3.mk_bvlshr(contextPtr, this.ast, this.sort.cast(count).ast)));
            }
            shl(count) {
                return new BitVecImpl(check(Z3.mk_bvshl(contextPtr, this.ast, this.sort.cast(count).ast)));
            }
            rotateRight(count) {
                return new BitVecImpl(check(Z3.mk_ext_rotate_right(contextPtr, this.ast, this.sort.cast(count).ast)));
            }
            rotateLeft(count) {
                return new BitVecImpl(check(Z3.mk_ext_rotate_left(contextPtr, this.ast, this.sort.cast(count).ast)));
            }
            not() {
                return new BitVecImpl(check(Z3.mk_bvnot(contextPtr, this.ast)));
            }
            extract(high, low) {
                return new BitVecImpl(check(Z3.mk_extract(contextPtr, high, low, this.ast)));
            }
            signExt(count) {
                return new BitVecImpl(check(Z3.mk_sign_ext(contextPtr, count, this.ast)));
            }
            zeroExt(count) {
                return new BitVecImpl(check(Z3.mk_zero_ext(contextPtr, count, this.ast)));
            }
            repeat(count) {
                return new BitVecImpl(check(Z3.mk_repeat(contextPtr, count, this.ast)));
            }
            sle(other) {
                return new BoolImpl(check(Z3.mk_bvsle(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            ule(other) {
                return new BoolImpl(check(Z3.mk_bvule(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            slt(other) {
                return new BoolImpl(check(Z3.mk_bvslt(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            ult(other) {
                return new BoolImpl(check(Z3.mk_bvult(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            sge(other) {
                return new BoolImpl(check(Z3.mk_bvsge(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            uge(other) {
                return new BoolImpl(check(Z3.mk_bvuge(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            sgt(other) {
                return new BoolImpl(check(Z3.mk_bvsgt(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            ugt(other) {
                return new BoolImpl(check(Z3.mk_bvugt(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            redAnd() {
                return new BitVecImpl(check(Z3.mk_bvredand(contextPtr, this.ast)));
            }
            redOr() {
                return new BitVecImpl(check(Z3.mk_bvredor(contextPtr, this.ast)));
            }
            addNoOverflow(other, isSigned) {
                return new BoolImpl(check(Z3.mk_bvadd_no_overflow(contextPtr, this.ast, this.sort.cast(other).ast, isSigned)));
            }
            addNoUnderflow(other) {
                return new BoolImpl(check(Z3.mk_bvadd_no_underflow(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            subNoOverflow(other) {
                return new BoolImpl(check(Z3.mk_bvsub_no_overflow(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            subNoUndeflow(other, isSigned) {
                return new BoolImpl(check(Z3.mk_bvsub_no_underflow(contextPtr, this.ast, this.sort.cast(other).ast, isSigned)));
            }
            sdivNoOverflow(other) {
                return new BoolImpl(check(Z3.mk_bvsdiv_no_overflow(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            mulNoOverflow(other, isSigned) {
                return new BoolImpl(check(Z3.mk_bvmul_no_overflow(contextPtr, this.ast, this.sort.cast(other).ast, isSigned)));
            }
            mulNoUndeflow(other) {
                return new BoolImpl(check(Z3.mk_bvmul_no_underflow(contextPtr, this.ast, this.sort.cast(other).ast)));
            }
            negNoOverflow() {
                return new BoolImpl(check(Z3.mk_bvneg_no_overflow(contextPtr, this.ast)));
            }
        }
        class BitVecNumImpl extends BitVecImpl {
            value() {
                return BigInt(this.asString());
            }
            asSignedValue() {
                let val = this.value();
                const size = BigInt(this.size());
                if (val >= 2n ** (size - 1n)) {
                    val = val - 2n ** size;
                }
                if (val < (-2n) ** (size - 1n)) {
                    val = val + 2n ** size;
                }
                return val;
            }
            asString() {
                return Z3.get_numeral_string(contextPtr, this.ast);
            }
            asBinaryString() {
                return Z3.get_numeral_binary_string(contextPtr, this.ast);
            }
        }
        class AstVectorImpl {
            constructor(ptr = Z3.mk_ast_vector(contextPtr)) {
                this.ptr = ptr;
                this.ctx = ctx;
                Z3.ast_vector_inc_ref(contextPtr, ptr);
                cleanup.register(this, () => Z3.ast_vector_dec_ref(contextPtr, ptr));
            }
            length() {
                return Z3.ast_vector_size(contextPtr, this.ptr);
            }
            [Symbol.iterator]() {
                return this.values();
            }
            *entries() {
                const length = this.length();
                for (let i = 0; i < length; i++) {
                    yield [i, this.get(i)];
                }
            }
            *keys() {
                for (let [key] of this.entries()) {
                    yield key;
                }
            }
            *values() {
                for (let [, value] of this.entries()) {
                    yield value;
                }
            }
            get(from, to) {
                const length = this.length();
                if (from < 0) {
                    from += length;
                }
                if (from >= length) {
                    throw new RangeError(`expected from index ${from} to be less than length ${length}`);
                }
                if (to === undefined) {
                    return _toAst(check(Z3.ast_vector_get(contextPtr, this.ptr, from)));
                }
                if (to < 0) {
                    to += length;
                }
                if (to >= length) {
                    throw new RangeError(`expected to index ${to} to be less than length ${length}`);
                }
                const result = [];
                for (let i = from; i < to; i++) {
                    result.push(_toAst(check(Z3.ast_vector_get(contextPtr, this.ptr, i))));
                }
                return result;
            }
            set(i, v) {
                _assertContext(v);
                if (i >= this.length()) {
                    throw new RangeError(`expected index ${i} to be less than length ${this.length()}`);
                }
                check(Z3.ast_vector_set(contextPtr, this.ptr, i, v.ast));
            }
            push(v) {
                _assertContext(v);
                check(Z3.ast_vector_push(contextPtr, this.ptr, v.ast));
            }
            resize(size) {
                check(Z3.ast_vector_resize(contextPtr, this.ptr, size));
            }
            has(v) {
                _assertContext(v);
                for (const item of this.values()) {
                    if (item.eqIdentity(v)) {
                        return true;
                    }
                }
                return false;
            }
            sexpr() {
                return check(Z3.ast_vector_to_string(contextPtr, this.ptr));
            }
        }
        class AstMapImpl {
            constructor(ptr = Z3.mk_ast_map(contextPtr)) {
                this.ptr = ptr;
                this.ctx = ctx;
                Z3.ast_map_inc_ref(contextPtr, ptr);
                cleanup.register(this, () => Z3.ast_map_dec_ref(contextPtr, ptr));
            }
            [Symbol.iterator]() {
                return this.entries();
            }
            get size() {
                return Z3.ast_map_size(contextPtr, this.ptr);
            }
            *entries() {
                for (const key of this.keys()) {
                    yield [key, this.get(key)];
                }
            }
            keys() {
                return new AstVectorImpl(Z3.ast_map_keys(contextPtr, this.ptr));
            }
            *values() {
                for (const [_, value] of this.entries()) {
                    yield value;
                }
            }
            get(key) {
                return _toAst(check(Z3.ast_map_find(contextPtr, this.ptr, key.ast)));
            }
            set(key, value) {
                check(Z3.ast_map_insert(contextPtr, this.ptr, key.ast, value.ast));
            }
            delete(key) {
                check(Z3.ast_map_erase(contextPtr, this.ptr, key.ast));
            }
            clear() {
                check(Z3.ast_map_reset(contextPtr, this.ptr));
            }
            has(key) {
                return check(Z3.ast_map_contains(contextPtr, this.ptr, key.ast));
            }
            sexpr() {
                return check(Z3.ast_map_to_string(contextPtr, this.ptr));
            }
        }
        const ctx = {
            ptr: contextPtr,
            name,
            /////////////
            // Classes //
            /////////////
            Solver: SolverImpl,
            Model: ModelImpl,
            Tactic: TacticImpl,
            AstVector: AstVectorImpl,
            AstMap: AstMapImpl,
            ///////////////
            // Functions //
            ///////////////
            interrupt,
            isModel,
            isAst,
            isSort,
            isFuncDecl,
            isApp,
            isConst,
            isExpr,
            isVar,
            isAppOf,
            isBool,
            isTrue,
            isFalse,
            isAnd,
            isOr,
            isImplies,
            isNot,
            isEq,
            isDistinct,
            isArith,
            isArithSort,
            isInt,
            isIntVal,
            isIntSort,
            isReal,
            isRealVal,
            isRealSort,
            isBitVecSort,
            isBitVec,
            isBitVecVal,
            isProbe,
            isTactic,
            isAstVector,
            eqIdentity,
            getVarIndex,
            from,
            solve,
            /////////////
            // Objects //
            /////////////
            Sort,
            Function,
            RecFunc,
            Bool,
            Int,
            Real,
            BitVec,
            ////////////////
            // Operations //
            ////////////////
            If,
            Distinct,
            Const,
            Consts,
            FreshConst,
            Var,
            Implies,
            Eq,
            Xor,
            Not,
            And,
            Or,
            ToReal,
            ToInt,
            IsInt,
            Sqrt,
            Cbrt,
            BV2Int,
            Int2BV,
            Concat,
            Cond,
        };
        cleanup.register(ctx, () => Z3.del_context(contextPtr));
        return ctx;
    }
    return {
        enableTrace,
        disableTrace,
        getVersion,
        getVersionString,
        getFullVersion,
        openLog,
        appendLog,
        getParam,
        setParam,
        resetParams,
        Context: createContext,
    };
}
exports.createApi = createApi;
